java/arcs/sdk/testing/BaseTestHarness.kt

```diff
@@ -91,53 +110,59 @@ open class BaseTestHarness<P : Particle>(
-                                spec,
-                                storageKey,
-                                immediateSync = false
-                            )
-                            // Harness (test) handle: allow full read/write access.
-                            harnessHandles[spec.baseName] = handleManager.createHandle(
-                                spec.copy(mode = HandleMode.ReadWrite),
-                                storageKey,
-                                immediateSync = false
-                            )
-                        }
-                    }
+                withSetupAndCleanup {
                     statement.evaluate()
-                    runBlocking {
-                        scheduler.waitForIdle()
-                        handleManager.close()
+                }
+            }
+        }
+    }
+
+    fun withSetupAndCleanup(run: () -> Unit) {
+        RamDisk.clear()
+        RamDiskDriverProvider()
+        DriverAndKeyConfigurator.configureKeyParsers()
+
+        val schedulerProvider = SimpleSchedulerProvider(Dispatchers.Default)
+        scheduler = schedulerProvider("testArc_${this.javaClass.simpleName}")
+        val handleManager = EntityHandleManager(
+            arcId = "testHarness",
+            hostId = "testHarnessHost",
+            time = JvmTime,
+            scheduler = scheduler,
+            storageEndpointManager = DirectStorageEndpointManager(StoreManager())
+        )
+        try {
+            runBlocking {
+                specs.forEach { spec ->
+                    val storageKey = when (spec.dataType) {
+                        HandleDataType.Entity -> ReferenceModeStorageKey(
+                            backingKey = RamDiskStorageKey("backing_${spec.baseName}"),
+                            storageKey = RamDiskStorageKey("entity_${spec.baseName}")
+                        )
+                        HandleDataType.Reference ->
+                            RamDiskStorageKey("ref_${spec.baseName}")
                     }
-                } finally {
-                    schedulerProvider.cancelAll()
+                    // Particle handle: use the manifest-specified read/write access.
+                    particleHandles[spec.baseName] = handleManager.createHandle(
+                        spec,
+                        storageKey,
+                        immediateSync = false
+                    )
+                    // Harness (test) handle: allow full read/write access.
+                    harnessHandles[spec.baseName] = handleManager.createHandle(
+                        spec.copy(mode = HandleMode.ReadWrite),
+                        storageKey,
+                        immediateSync = false
+                    )
                 }
             }
+            run()
+            runBlocking {
+                scheduler.waitForIdle()
+                handleManager.close()
```

Comment:
shouldn't the handleManager.close() be in the finally below?

Reply:
I think so. Moved.